### PR TITLE
feat(traefik,consul-catalog): Set attributes from consul catalog tags (loadbalancer,circuitbreaker,weight)

### PR DESCRIPTION
**Pull request by samber**
see #254 #253

### CODE REVIEW

File: templates/consul_catalog.tmpl [hunk 1]
```diff
@@ -1,12 +1,33 @@
-[backends]{{range .Nodes}}
+[backends]
+{{range .Nodes}}
     [backends.backend-{{getBackend .}}.servers.server-{{.Node.Node | replace "." "-"}}-{{.Service.Port}}]
-    url = "http://{{.Node.Address}}:{{.Service.Port}}"
+    	url = "http://{{.Node.Address}}:{{.Service.Port}}"
```
Comment (tilgovi) [inline templates/consul_catalog.tmpl]: Would it be easy to add `traefik.protocol` here as a way to support HTTPS to backends from consul?

Comment (tilgovi) [inline templates/consul_catalog.tmpl]: Maybe just this does the trick (waiting for build and setting up tests over here):

    diff --git a/templates/consul_catalog.tmpl b/templates/consul_catalog.tmpl
    index 4e86f32..1cdbeeb 100644
    --- a/templates/consul_catalog.tmpl
    +++ b/templates/consul_catalog.tmpl
    @@ -1,7 +1,12 @@
     [backends]
     {{range .Nodes}}
         [backends.backend-{{getBackend .}}.servers.server-{{.Node.Node | replace "." "-"}}-{{.Service.Port}}]
    +      {{$scheme := getAttribute "protocol" .Service.Tags}}
    +      {{if $scheme}}
    +      url = "{{$scheme}}//{{.Node.Address}}:{{.Service.Port}}"
    +      {{else}}
         	url = "http://{{.Node.Address}}:{{.Service.Port}}"
    +      {{end}}
             {{$weight := getAttribute "weight" .Service.Tags}}
             {{with $weight}}
                 weight = {{$weight}}
    @@ -30,4 +35,4 @@
       passHostHeader = false
         [frontends.frontend-{{.ServiceName}}.routes.route-host-{{.ServiceName}}]
         rule = "{{getFrontendValue .ServiceName}}"
    -{{end}}
    \ No newline at end of file
    +{{end}}

File: templates/consul_catalog.tmpl [hunk 2]
```diff
@@ -1,12 +1,33 @@
...
+    {{$circuitBreaker := getAttribute "circuitbreaker" .Attributes}}
+    {{with $circuitBreaker}}
+    [backends.backend-{{$service}}.circuitbreaker]
+        expression = "{{$circuitBreaker}}"
+    {{end}}
+
+    {{$loadBalancer := getAttribute "loadbalancer" .Attributes}}
+    {{with $loadBalancer}}
+    [backends.backend-{{$service}}.loadbalancer]
+    	method = "{{$loadBalancer}}"
```
Comment (tilgovi) [inline templates/consul_catalog.tmpl]: This line contains a tab, inconsistent with the spaces used in the rest of the file.

File: templates/consul_catalog.tmpl (hunk 1 above)
Comment (tilgovi) [inline templates/consul_catalog.tmpl]: Also, this line contains a tab, inconsistent with the spaces used in the rest of the file.

**Comment by emilevauge** [comment]
Thanks @samber for your contribution!
Thanks @tilgovi for your review ;)

Don't you think it would be great to also include other tags?

- `traefik.backend`
- `traefik.protocol`
- `traefik.enable`
- `traefik.frontend.rule`
- `traefik.frontend.passHostHeader`
- `traefik.frontend.entryPoints`
- `traefik.domain`

**Comment by tilgovi** [comment]
I think it should be possible to implement all these, but some more processing needs to be done in `buildConfig` to aggregate the backends and frontends to pass to the template.

**Comment by tilgovi** [comment]
@samber I implemented `traefik.protocol` and `traefik.enable` quickly here: https://github.com/tilgovi/traefik/commit/f599fb530ed6d1bdff287107d53862e8d6fa96ef

I based that on your work, specifically the key/value extraction, but in thinking about the other attributes I came to the conclusion above, that we should process this all a bit more in `buildConfig`. I needed this to work right now, so I stripped your change down to what I have there, but I'm happy to leave this with you or help out as I can.

File: provider/consul_catalog.go [hunk 3]
```diff
@@ -79,8 +86,11 @@ func (provider *ConsulCatalog) healthyNodes(service string) (catalogUpdate, erro
 	}
 
 	return catalogUpdate{
-		Service: service,
-		Nodes:   data,
+		Service: &serviceUpdate{
+			ServiceName: service,
+			Attributes:  data[0].Service.Tags,
```
Comment (keis) [inline provider/consul_catalog.go]: tags are also available through `/services` aka `watchServices` where they are a union of all tags for the service. I'm not sure if that's in any way better than picking the first but it's perhaps a bit more consistent.

**Comment by samber** [comment]
@keis: sure, it is much more consistant, but in return /v1/catalog/services give you nodes that does not pass health check.
The trade-off could be to put every tags in a single set (healthyNodes function) after the healthcheck. You will be able to use a tag from the template even if a it is not set for node[0]

@tilgovi: getAttribute now accept a default value

@emilevauge:
* ```traefik.backend```: what do you expect ?
* ```traefik.protocol```: ok
* ```traefik.enable```: ok
* ```traefik.frontend.rule```: ok
* ```traefik.frontend.passHostHeader```: ok
* ```traefik.frontend.entryPoints```: WIP
* ```traefik.domain```: see ```traefik.frontend.rule```

I renamed attributes names to be more consistant with consul-kv:
- ```traefik.protocol``` => ```traffic.backend.protocol```
- ```traefik.enable``` => ```traffic.backend.enable```
- ```traefik.weight``` => ```traffic.backend.weight```
- ```traefik.loadbalancer``` => ```traffic.backend.loadbalancer```
- ```traefik.circuitbreaker``` => ```traffic.backend.circuitbreaker```

All: K/V store in service catalog is coming with Consul 0.8

I expect to commit doc and tests tomorrow.

**Comment by tilgovi** [comment]
Does this seem correct?

- Start with an empty set of frontends and backends
- For each (service, node) pair...
  - inspect `traefik.domain` or `traefik.frontend` tags
  - Create the appropriate frontend if no such frontend exists, and its corresponding backend
  - Inspect `traefik.backend` tag (default to the service name)
  - Create the appropriate backend if no such backend exists
  - Add the node (protocol, address, port) to the backend as a server

**Comment by keis** [comment]
@samber I was talking about the list of services `/v1/catalog/services` (which is already being used) and not the instance list of a service `/v1/catalog/service/foo`. But looking at each node is perhaps the better solution anyway. 

@tilgovi 
You did not mention it but I guess `traefik.frontend` should also default to the service name?
What happens if we have conflicting values for any of the tags? like 'frontend.passHostHeader` or `backend.loadbalancer`

**Comment by samber** [comment]
@keis we call /v1/catalog/services, then /v1/health/service/<service> for each service
Tags come from healthy services ;-)

@emilevauge @tilgovi Are you talking about setting a custom frontend/backend name ? Can you tell me more about the use case ?

**Comment by samber** [comment]
Sorry, I just see the k/v format of other backeds. Let me update ;-)

**Comment by samber** [comment]
Ready for review ;-)

File: provider/consul_catalog.go [hunk 4]
```diff
@@ -16,6 +16,8 @@ import (
 const (
 	// DefaultWatchWaitTime is the duration to wait when polling consul
 	DefaultWatchWaitTime = 15 * time.Second
+	// ConsulCatalogTagPrefix is a prefix for additional service/node configurations
+	ConsulCatalogTagPrefix = "traefik."
```
Comment (emilevauge) [inline provider/consul_catalog.go]: I would instead add a `Prefix` field in struct `ConsulCatalog`, like in https://github.com/containous/traefik/blob/master/provider/kv.go#L25.

File: provider/consul_catalog.go [hunk 5]
```diff
@@ -79,41 +86,76 @@ func (provider *ConsulCatalog) healthyNodes(service string) (catalogUpdate, erro
...
+	return strings.Split(list, ",")
+}
+
 func (provider *ConsulCatalog) getBackend(node *api.ServiceEntry) string {
 	return strings.ToLower(node.Service.Service)
 }
 
-func (provider *ConsulCatalog) getFrontendValue(service string) string {
-	return "Host:" + service + "." + provider.Domain
+func (provider *ConsulCatalog) getFrontendValue(service serviceUpdate) string {
```
Comment (emilevauge) [inline provider/consul_catalog.go]: Could you rename it to getFrontendRule ?

**Comment by emilevauge** [comment]
Thanks @samber :)
I made minor comments.
Could you also rebase and squash your commits ?

**Comment by emilevauge** [comment]
@samber, sorry it needs a rebase again :angel: 
LGTM
/cc @vdemeester

**Comment by vdemeester** [comment]
LGTM :+1:

**Comment by tilgovi** [comment]
Excellent!

**Pull request merged**